File: Day337_2025/Q337_GFG.java
class Q337_GFG {
    public int tsp(int[][] cost) {
        int n = cost.length;
        int N = 1 << n;

        int[][] dp = new int[N][n];
        int INF = (int)1e9;

        // Initialize dp with INF
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < n; j++) {
                dp[i][j] = INF;
            }
        }

        dp[1][0] = 0;  // Starting at city 0

        for (int mask = 0; mask < N; mask++) {
            for (int u = 0; u < n; u++) {
                if ((mask & (1 << u)) == 0) continue;

                for (int v = 0; v < n; v++) {
                    if ((mask & (1 << v)) != 0) continue;

                    int nextMask = mask | (1 << v);

                    dp[nextMask][v] = Math.min(dp[nextMask][v],
                                               dp[mask][u] + cost[u][v]);
                }
            }
        }

        int ans = INF;
        int fullMask = (1 << n) - 1;

        for (int u = 0; u < n; u++) {
            ans = Math.min(ans, dp[fullMask][u] + cost[u][0]);
        }

        return ans;
    }
}

File: Day337_2025/Q337_GFG.py
class Solution:
    def tsp(self, cost):
        n = len(cost)
        N = 1 << n
        INF = 10**9

        dp = [[INF] * n for _ in range(N)]
        dp[1][0] = 0  # Start at city 0

        for mask in range(N):
            for u in range(n):
                if not (mask & (1 << u)):
                    continue
                for v in range(n):
                    if mask & (1 << v):
                        continue
                    nextMask = mask | (1 << v)
                    dp[nextMask][v] = min(dp[nextMask][v],
                                          dp[mask][u] + cost[u][v])

        ans = INF
        fullMask = (1 << n) - 1

        for u in range(n):
            ans = min(ans, dp[fullMask][u] + cost[u][0])

        return ans
